File: core/src/com/xenoage/zong/core/format/SystemLayout.java
package com.xenoage.zong.core.format;

import static com.xenoage.utils.collections.CollectionUtils.setExtend;

import java.util.List;

import lombok.Data;

import com.xenoage.utils.annotations.MaybeNull;

/**
 * Layout information for a system.
 *
 * @author Andreas Wenger
 */
@Data
public final class SystemLayout {

	public static final float defaultDistance = 30f;
	public static final float defaultMarginLeft = 5f;
	public static final float defaultMarginRight = 5f;

	/** The distance between the bottom line
	 * of the previous system to the top line of this system in mm.
	 * If this is the first system of a frame, this is the distance
	 * to the top margin of the frame (attention: this is different
	 * in MusicXML, where a special top-system-distance is used). */
	private float distance = defaultDistance;
	/** The distance between the left side of the staves and the left page margin. */
	private float marginLeft = defaultMarginLeft;
	/** The distance between the right side of the staves and the left page margin. */
	private float marginRight = defaultMarginRight;
	/** The layouts of the staves of the system (may also be or contain null). */
	@MaybeNull public List<StaffLayout> staffLayouts = null;


	/**
	 * Gets layout information for the staff with the given index,
	 * or null if undefined.
	 */
	public StaffLayout getStaffLayout(int staffIndex) {
		if (staffLayouts != null && staffIndex >= 0 && staffIndex < staffLayouts.size())
			return staffLayouts.get(staffIndex);
		else
			return null;
	}

	/**
	 * Sets the {@link StaffLayout} for the at staff given index.
	 */
	public void setStaffLayout(int staffIndex, StaffLayout staffLayout) {
		staffLayouts = setExtend(staffLayouts, staffIndex, staffLayout, null);
	}

}

File: core/src/com/xenoage/zong/core/format/SystemLayout.kt
package com.xenoage.zong.core.format

import com.xenoage.utils.Mm
import com.xenoage.utils.setExtend

/**
 * Layout information for a system.
 */
data class SystemLayout(
	/** The distance between the bottom line
	 * of the previous system to the top line of this system in mm.
	 * If this is the first system of a frame, this is the distance
	 * to the top margin of the frame (attention: this is different
	 * in MusicXML, where a special top-system-distance is used).  */
	var distance: Mm = defaultDistance,
	/** The distance between the left side of the staves and the left page margin */
	var marginLeft: Mm = defaultMarginLeft,
	/** The distance between the right side of the staves and the left page margin */
	var marginRight: Mm = defaultMarginRight,
	/** The layouts of the staves of the system (may also contain null).  */
	val staffLayouts: MutableList<StaffLayout?> = mutableListOf()
) {

	/**
	 * Gets layout information for the staff with the given index,
	 * or null if undefined.
	 */
	fun getStaffLayout(staffIndex: Int): StaffLayout? =
			staffLayouts?.getOrNull(staffIndex)

	/**
	 * Sets the [StaffLayout] for the at staff given index.
	 */
	fun setStaffLayout(staffIndex: Int, staffLayout: StaffLayout) {
		staffLayouts.setExtend(staffIndex, staffLayout, null)
	}

	companion object {
		val defaultDistance: Mm = 30f
		val defaultMarginLeft: Mm = 5f
		val defaultMarginRight: Mm = 5f
	}

}
